File: src/main/java/kt/scalerdesigns/Builder.java
package kt.scalerdesigns;

public class Builder {
    public static void main(String[] args) {
        Student student = Student.getBuilder().setName("la")
                .setAge("45")
                .build();
    }
}

class Student {
    private String name;
    private String age;
    private Student(Builder builder) {
        name = builder.name;
        age = builder.age;
    }

    static Builder getBuilder() {
        return new Builder();
    }
    static class Builder {
        private Builder() {}
        private String name;
        private String age;

        Builder setName(String name) {
            this.name = name;
            return this;
        }

        Builder setAge(String age) {
            this.age = age;
            return this;
        }

        Student build() {
            return new Student(this);
        }
    }
}

File: src/main/java/kt/scalerdesigns/Builder.kt
//package kt.scalerdesigns
//
//import kt.academy.learnprogramming.challenge2.delegation.name
//import kt.generics.doSomething
//
//class Student private constructor(val builder: Builder) {
//    private var name: String? = null
//    private var age: Int? = null
//    private var height: Float? = null
//
//    init {
//        name = builder.name
//        age = builder.age
//        height = builder.height
//    }
//
//
//     class Builder private constructor() {
//        companion object {
//            fun getBuilder() = Builder()
//        }
//        var name: String? = null
//        var age: Int? = null
//        var height: Float? = null
//
//        fun setName(name: String) = apply { this@Builder.name = name }
//        fun setAge(age: Int) = apply { this@Builder.age = age }
//        fun setHeight(height: Float) = apply { this@Builder.height = height }
//        fun build() = Student(this)
//    }
//}
//
//fun main() {
//    val student: Student = Student.Builder.getBuilder().setAge(23)
//            .setHeight(34.0f)
//            .setName("lal")
//            .build()
//}